luaui/Widgets/unit_factory_quota.lua

```diff
@@ -0,0 +1,253 @@
+    end
+end
+
+local function fillQuotas()
+    for factID, quota in pairs(quotas) do
+        if isFactoryUsable(factID) then
+            for udefid, num in pairs(quota) do
+                if num == 0 then
+                    quota[udefid] = nil
+                end
+            end
+            if table.length(quota)>0 then
+                local function isBetter(q1, q2, k1, k2)
+                    return (table.length((builtUnits[factID] or {})[k1] or {})/q1) < (table.length((builtUnits[factID] or {})[k2] or {})/q2)
+                end
+                local quotaNum, uDefID = findMin(quota, isBetter)
+                if quotaNum > table.length((builtUnits[factID] or {})[uDefID] or {}) then
+                    insertToFactQ(factID, -uDefID, {}, {"alt"})
+                end
+            end
+        end
+    end
+end
+
+function widget:GameFrame(n)
+    if n % 4 == 0 then -- improve perfomance
+        fillQuotas()
+    end
+end
+
+----- handle resetting queue mode to normal
+local function isOnQuotaBuildMode(unitID)
+	return GetUnitCmdDescs(unitID)[FindUnitCmdDesc(unitID, CMD_QUOTA_BUILD_TOGGLE)].params[1]+0 == 1
+end
+
+function widget:SelectionChanged(newSel)
+    for _, unitID in ipairs(newSel) do
+        if factoryDefIDs[GetUnitDefID(unitID)] and isOnQuotaBuildMode(unitID) then
+            GiveOrderToUnit(unitID, CMD_QUOTA_BUILD_TOGGLE, {0}, {})
+        end
+    end
+end
+
+----- handle unit tracking
+function widget:UnitCreated(unitID, unitDefID, unitTeam, builderID)
+    if builderID and unitTeam == myTeam and factoryDefIDs[GetUnitDefID(builderID)] then
+        builtUnits[builderID] = builtUnits[builderID] or {}
+        builtUnits[builderID][unitDefID] = builtUnits[builderID][unitDefID] or {}
+        builtUnits[builderID][unitDefID][unitID] = true
+        unitFacts[unitID] = builderID
+    end
+end
+
+local function removeUnit(unitID, unitDefID, unitTeam)
+    if unitTeam == myTeam and unitFacts[unitID] then --check if it was built by the same player
+        builtUnits[unitFacts[unitID]][unitDefID][unitID] = nil
+        unitFacts[unitID] = nil
+    end
+end
+
+function widget:UnitDestroyed(unitID, unitDefID, unitTeam)
+    removeUnit(unitID, unitDefID, unitTeam)
+end
+
+function widget:UnitGiven(unitID, unitDefID, newTeam, oldTeam)
+    if myTeam == oldTeam and myTeam ~= newTeam then
+        removeUnit(unitID, unitDefID, myTeam)
+    end
+end
+
+
+
+function widget:PlayerChanged(playerID)
+    if Spring.GetSpectatingState() then
+        widgetHandler:RemoveWidget(self)
+    end
+    myTeam = Spring.GetMyTeamID()
+end
+
+function widget:Initialize()
+    widget:PlayerChanged()
+
+    for unitName, _ in pairs(quotas) do
+        quotas[UnitDefNames[unitName].id] = quotas[unitName]
+        quotas[unitName] = nil
+    end
+
+    WG.Quotas = {}
+    WG.Quotas.getQuotas = function()
+        return quotas
+    end
+    WG.Quotas.update = function(newQuotas)
+        quotas = newQuotas
+    end
```

Comment:
Wouldn't it be more efficient to have the update function be passed a queue ID and update the single queue, rather than having to update the entire table every time?

Reply:
I don't even know if this is necessary, since the table is never copied in the other widgets, so I even think calling this might be unnecessary.

Reply:
I use this update in buildings_and_units_tracker, pls don't remove :D

Reply:
> I use this update in buildings_and_units_tracker, pls don't remove :D

```lua
local quotas = WG.Quotas.getQuotas()
quotas = changedQuotas
```
This should do the same thing as `WG.Quotas.update(changedQuotas)`

Reply:
Would it be possible if I remove it from all other widgets, but keep it exposed for backwards compatibility?

Reply:
> I use this update in buildings_and_units_tracker, pls don't remove :D

How could you be using this functionality when it's not merged yet?